File: rust/src/atcoder/beginner/q110/d.rs
use std;

pub fn answer() {
    fn read<T: std::str::FromStr>() -> T {
        let mut s = String::new();
        std::io::stdin().read_line(&mut s).ok();
        s.trim().parse().ok().unwrap()
    }
    fn read_vec<T: std::str::FromStr>() -> Vec<T> {
        read::<String>().split_whitespace().map(|e| e.parse().ok().unwrap()).collect()
    }

    fn primes(max:i64) -> Vec<i64>{
        let mut k = vec![2,3];
        for c in (1..(max/6)+1){
            let a = c*6-1;
            let b = c*6+1;
            let mut af = true;
            let mut bf = true;
            for d in (0..k.len()) {
                if(af && (a % k[d]) == 0){
                    af = false;
                }
                if(bf && (b % k[d]) == 0){
                    bf = false;
                }
            }
            if(af){
                k.append(&mut vec![a])
            }
            if(bf){
                k.append(&mut vec![b])
            }
        }
        return k
    }
    fn invmod(p:i64, g:i64) -> i64 {
        fn euclidean(px:i64, py:i64, pgcd:i64, x:i64, y:i64, gcd:i64) -> i64 {
            match (pgcd % gcd) {
                0 => x,
                _ => {
                    let d = pgcd /gcd;
                    let m = pgcd % gcd;
                    return euclidean(x, y, gcd, px-(d*x), py-(d*y), m)
                }
            }
        }
        return (euclidean(1,0,p,0,1,g)+g)%g;
    }
    fn multiper(n:i64, m:i64, c:i64) -> (i64, i64) {
        match m % n {
            0 => multiper(n, m/n, c+1),
            _ => (m,c)
        }
    }

    fn comb(n:i64, k:i64, p:i64)-> i64 {
        let mut ans = 1;
        for c in (1..k+1){
            ans = remain(multi(ans , (n-c+1)), p);
            ans = remain(multi(ans, invmod(c, p)), p);
//            ans = remain(divide(ans , c), p);
        }
        return ans
    }

    fn fac(l:Vec<i64>, m:i64) -> Vec<i64> {
        match l.len() {
            0 => vec![1],
            _ => {
                let mut ll = l;
                let a = match multiper(ll.pop().unwrap(), m, 0) {
                    (1, 0) => vec![],
                    (1, y) => vec![y],
                    (x, 0) => {
                        return fac(ll, m)
                    }
                    (x, y) => {
                        let mut k = vec![y];
                        k.append(&mut fac(ll,x));
                        return k
                    }
                };
                return a;
            }
        }
    }

    fn multi(a:i64, b:i64) -> i64 {a*b}
    fn divide(a:i64, b:i64) -> i64 {a/b}
    fn remain(a:i64, b:i64) -> i64 {a%b}
    fn main() {
        let max = 1000000007;
        let mut nk = read_vec::<i64>();
        let n:i64 = nk[0];
        let m:i64 = nk[1];
        let primes = primes(((m as f64).sqrt() as i64));

        let f = fac(primes, m);
        let mut a = 1;
        for c in f {
            a = remain(multi(a,comb(n+c-1, c, max)),max)
        }
        println!("{}", a);
    }
    main()
}

File: scala/src/main/scala/atcoder/beginner/q110/D.scala
package atcoder.beginner.q110

import scala.io.StdIn.readLine

object D {
  def multi(a:Long, b:Long):Long = a * b
  def divide(a: Long, b: Long):Long = a / b
  def remain(a: Long, b: Long):Long = a % b

  def primes(max:Int): List[Int] = {
    (1 to (max / 6).toInt + 1).foldLeft(List(2,3)){(acc,c) => {
      val a = c*6-1
      val b = c*6+1
      val sq = acc.takeWhile(a=> a <= math.sqrt(b).toInt+1)
      val k=if(sq.foldLeft(true){(acc, c) => if(acc && remain(a, c) != 0) true else false}) List(a) else List()
      val v=if(sq.foldLeft(true){(acc, c) => if(acc && remain(b, c) != 0) true else false}) List(b) else List()
      acc++k++v
    }}
  }
  def multiper(n:Int, m:Int, c:Int):(Int, Int) = if(m % n == 0) multiper(n, m/n, c+1) else (c,m)

  def fac(l:List[Int], m:Int): List[Int] = {
    l match {
      case List() => List(1) // math.sqrt(math.pow(10,9).toInt) < math.pow(10,9)の素数が入っていることを想定
      case _ => multiper(l.head, m, 0) match {
        case (0,1) => List()
        case (x,1) => List(x)
        case (0,y) => fac(l.tail, y)
        case (x,y) => List(x) ::: fac(l.tail, y)
      }
    }
  }
  def invmod(p:Long, g:Long) = {
    def euclidean(px:Long, py:Long, pgcd: Long, x:Long, y:Long, gcd: Long):Long = {
      pgcd % gcd match {
        case 0 => (x + g) % g
        case m => {
          val d = pgcd / gcd
          euclidean(x,y,gcd, px-(d*x), py-(d*y), m)
        }
      }
    }
    euclidean(1,0,p,0,1,g)
  }
  def comb(n:Long, k:Long, p:Long):Long = {
    (1 to k.toInt).foldLeft(1.toLong)((acc,c)=> remain((multi(remain(multi(acc,(n-c+1)), p), invmod(c,p))), p))
  }
  def input[A](change: String => A) = readLine.split(" ").map(change(_))
  def main(args: Array[String]): Unit ={
    val p = math.pow(10,9).toLong + 7
    println(comb(100008,1,p))
    println(comb(100008,2,p))
    println(comb(100008,3,p))
    println(comb(100008,4,p))
    println(comb(100008,5,p))
    println(comb(100008,6,p))
    println(comb(100008,7,p))
    println(comb(100008,8,p))
    println(comb(100008,9,p))
    //    val ps = primes(math.sqrt(math.pow(10,9).toInt).toInt)
    //    val a = input((a=>a.toInt))
    //    val n = a(0)
    //    val m = a(1)
    //    val f = fac(primes(math.sqrt(p).toInt), m)
    //    val ans = f.foldLeft(1.toLong)((acc,c)=>(multi(acc, comb(c+n-1, c, p))) % p)
    //    println(ans)
  }
}
